server/service/src/invoice/outbound_return/insert/mod.rs

```diff
@@ -0,0 +1,367 @@
+use repository::{
+    ActivityLogType, Invoice, InvoiceRowRepository, RepositoryError, TransactionError,
+};
+
+use crate::{
+    activity_log::activity_log_entry,
+    invoice::get_invoice,
+    invoice_line::{
+        stock_out_line::{insert_stock_out_line, InsertStockOutLineError},
+        update_return_reason_id::{update_return_reason_id, UpdateLineReturnReasonError},
+    },
+    service_provider::ServiceContext,
+};
+pub mod generate;
+pub mod validate;
+use generate::generate;
+use validate::validate;
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct InsertOutboundReturn {
+    pub id: String,
+    pub other_party_id: String,
+    pub outbound_return_lines: Vec<InsertOutboundReturnLine>,
+}
+
+#[derive(Clone, Debug, Default, PartialEq)]
+pub struct InsertOutboundReturnLine {
+    pub id: String,
+    pub stock_line_id: String,
+    pub number_of_packs: f64,
+    pub reason_id: Option<String>,
+    pub note: Option<String>,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum InsertOutboundReturnError {
+    InvoiceAlreadyExists,
+    // Name validation
+    OtherPartyNotASupplier,
+    OtherPartyNotVisible,
+    OtherPartyDoesNotExist,
+    // Internal
+    NewlyCreatedInvoiceDoesNotExist,
+    DatabaseError(RepositoryError),
+    // Line Errors
+    LineInsertError {
+        line_id: String,
+        error: InsertStockOutLineError,
+    },
+    LineReturnReasonUpdateError {
+        line_id: String,
+        error: UpdateLineReturnReasonError,
+    },
+}
+
+type OutError = InsertOutboundReturnError;
+
+pub fn insert_outbound_return(
+    ctx: &ServiceContext,
+    input: InsertOutboundReturn,
+) -> Result<Invoice, OutError> {
+    let outbound_return: Invoice = ctx
```

Comment:
I like how this reads now, I think easier to see the overall intention at the top level, and then it's up to the reader if they want to get bogged down with more detail, in the section they are interested in